Website/Providers/DataProviders/SqlDataProvider/09.02.02.SqlDataProvider

```diff
@@ -76,32 +76,56 @@ UPDATE {databaseOwner}{objectQualifier}Packages SET Version = @version WHERE Nam
+  DROP PROCEDURE {databaseOwner}[{objectQualifier}DeleteOrphanedAspNetUsers]
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}DeleteOrphanedAspNetUsers]
+    @ApplicationName NVARCHAR(256)
+AS
+
 IF (SELECT COUNT(*) FROM dbo.sysobjects
-	WHERE id = object_id(N'{databaseOwner}[{objectQualifier}aspnet_Users]') OR
-		  id = object_id(N'{databaseOwner}[{objectQualifier}aspnet_Membership]') OR
+	WHERE id = object_id(N'dbo.[aspnet_Users]') OR
+		  id = object_id(N'dbo.[aspnet_Membership]') OR
 		  id = object_id(N'{databaseOwner}[{objectQualifier}Users]') and OBJECTPROPERTY(id, N'IsUserTable') = 1) = 3
 BEGIN
 	BEGIN TRANSACTION
-		-- remove temp table if exists
-		IF OBJECT_ID('tempdb..#temp_orphaned_users') IS NOT NULL
-			DROP TABLE #temp_orphaned_users
-		-- prepare orphaned users to delete
-		SELECT aspUsers.UserId INTO #temp_orphaned_users FROM {databaseOwner}{objectQualifier}aspnet_Users aspUsers
-		LEFT JOIN {databaseOwner}{objectQualifier}Users users ON aspUsers.UserName = users.Username		
-		WHERE users.Username IS NULL
-		-- remove users from asp membership
-		DELETE m FROM {databaseOwner}{objectQualifier}aspnet_Membership m
-		INNER JOIN #temp_orphaned_users o ON m.UserId = o.UserId
-		-- remove users from asp users
-		DELETE u FROM {databaseOwner}{objectQualifier}aspnet_Users u
-		INNER JOIN #temp_orphaned_users o ON u.UserId = o.UserId
-		-- remove temp table if exists
-		IF OBJECT_ID('tempdb..#temp_orphaned_users') IS NOT NULL
-			DROP TABLE #temp_orphaned_users
-	COMMIT
+		BEGIN TRY
+			-- remove orphaned aspnet memberships
+			;WITH
+			orphaned_aspnet_memberships(UserId)
+			AS
+			(SELECT asp_memberships.UserId FROM dbo.aspnet_Membership asp_memberships
+					INNER JOIN dbo.aspnet_Applications asp_apps ON asp_apps.ApplicationId = asp_memberships.ApplicationId
+					INNER JOIN dbo.aspnet_Users asp_users ON asp_memberships.UserId = asp_users.UserId
+					LEFT JOIN {databaseOwner}{objectQualifier}Users users ON asp_users.UserName = users.Username
+					WHERE asp_apps.LoweredApplicationName = LOWER(@ApplicationName) AND users.Username IS NULL)
+			DELETE m FROM dbo.aspnet_Membership m
+			INNER JOIN orphaned_aspnet_memberships o ON m.UserId = o.UserId
+
+			-- remove orphaned aspnet users
+			;WITH
+			orphaned_aspnet_users (UserId)
+			AS 
+			(SELECT asp_users.UserId FROM dbo.aspnet_Users asp_users
+					INNER JOIN dbo.aspnet_Applications asp_apps ON asp_apps.ApplicationId = asp_users.ApplicationId
+					LEFT JOIN {databaseOwner}{objectQualifier}Users users ON asp_users.UserName = users.Username		
+					WHERE asp_apps.LoweredApplicationName = LOWER(@ApplicationName) AND users.Username IS NULL)
+			DELETE u FROM dbo.aspnet_Users u
+			INNER JOIN orphaned_aspnet_users o ON u.UserId = o.UserId
+		END TRY
+		BEGIN CATCH
+			SELECT ERROR_MESSAGE() AS ErrorMessage
+			IF @@TRANCOUNT > 0  
+				ROLLBACK TRANSACTION
+		END CATCH
+	IF @@TRANCOUNT > 0 
+	    COMMIT TRANSACTION
```

Comment:
Hm, if there were no errors, I think you should always commit. Otherwise won't the transaction be rolled back automatically when connection closes?

Reply:
@tpluscode 
The construction is following:
BEGIN TRANSACTION - TRY/CATCH - COMMIT
I'm always commit if there are no errors. If errors, transaction rolled back, so nothing to commit at the end.
Removing rollback command or doing commit in a try block will not change the essence of current implementation. All commands are called explicitly, so we know what happens on every step.

Reply:
You're right. I misunderstood what I was looking at. Thanks